includes/sanitizers/class-amp-href-sanitizer.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Class AMP_Href_Sanitizer.
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Href_Sanitizer.
+ *
+ * Crawls all href attributes to ensure their URLs are valid.
+ */
+final class AMP_Href_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Xpath query selector to target all elements with a href attribute.
+	 *
+	 * @var string
+	 */
+	const XPATH_SELECTOR = '//*[@href]';
```

Comment:
I think this selector may not be complete, since there are more attributes with URL values than `href`.

Essentially, this needs to apply to every attribute that has a `value_url` property in the [`AMP_Allowed_Tags_Generated`](https://github.com/ampproject/amp-wp/blob/75674dfab5ac7f0fb59c35828d2501326240bbbc/includes/sanitizers/class-amp-allowed-tags-generated.php), for example:

https://github.com/ampproject/amp-wp/blob/75674dfab5ac7f0fb59c35828d2501326240bbbc/includes/sanitizers/class-amp-allowed-tags-generated.php#L17466-L17474

There are many examples of this, such as `xlink:href`, `src`, `srcset`, `data-tooltip-icon`, `data-url`, `data-share-media`, `data-share-url`, `config`, `data-href`, `data-src`, `data-tag`, and so on. 

Therefore, I think the best place to put this code is actually going to be in [`AMP_Tag_And_Attribute_Sanitizer`](https://github.com/ampproject/amp-wp/blob/75674dfab5ac7f0fb59c35828d2501326240bbbc/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php), in particular somewhere that `AMP_Rule_Spec::VALUE_URL` is being looked for:

https://github.com/ampproject/amp-wp/blob/75674dfab5ac7f0fb59c35828d2501326240bbbc/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php#L837-L887

Then there won't be a need to manually maintain this list, and the validation would be done at a point where it is already walking over the tree.

Reply:
I'll try this in a separate PR again, to see what the result will be. We can later close whatever we don't want to use.

Reply:
I created a new PR #3151 with this approach. If that PR is preferable, we can close this one here.